persistent/Database/Persist/Sql/Migration.hs

```diff
@@ -294,13 +306,22 @@ newtype PersistUnsafeMigrationException
 instance Show PersistUnsafeMigrationException where
   show (PersistUnsafeMigrationException mig) =
     concat
-      [ "\n\nDatabase migration: manual intervention required.\n"
-      , "The unsafe actions are prefixed by '***' below:\n\n"
-      , unlines $ map displayMigration mig
+      [ "\n\nDatabase migration: manual intervention required.\n",
+        "The unsafe actions are prefixed by '***' below:\n\n",
+        unlines $ map displayMigration mig
       ]
     where
       displayMigration :: (Bool, Sql) -> String
-      displayMigration (True,  s) = "*** " ++ unpack s ++ ";"
+      displayMigration (True, s) = "*** " ++ unpack s ++ ";"
       displayMigration (False, s) = "    " ++ unpack s ++ ";"
 
 instance Exception PersistUnsafeMigrationException
+
+-- | Create an index for the given 'EntityField' and table.
+--
+-- @since 2.13.3.0
+createSearchIndex :: (PersistEntity rec) => EntityField rec typ -> Text -> Migration
+createSearchIndex entity tableName =
+  addMigration True $ "CREATE INDEX IF NOT EXISTS ON \"" <> tableName <> "\" (\"" <> fieldName <> "\")"
+  where
+    fieldName = unFieldNameDB . fieldDB $ persistFieldDef entity
```

Comment:
The only way I was able to get this to compile was to add a ``PersistEntity typ`` constraint, but this doesn't make sense as that's the raw value itself.

I've also tried to use ``entityDef`` on the ``fieldDef`` itself:
```haskell
entityDef $ persistFieldDef entityfield
```
But this ends up with the error `` Couldn't match expected type ‘proxy0  record0’``.

I know the ``PersistEntity typ`` constraint is incorrect, but couldn't figure out how to get the ``entityDef`` properly.
What am I missing here?

Reply:
The first argument to `entityDef` would be something like `entityDef :: Proxy rec -> EntityDef`.

First, we'd want to bring the type into scope, so we can refer to `rec`. This means we may need to add `ScopedTypeVariables` and then write an explicit `forall`:

```haskell
createSearchIndex :: forall rec typ. (PersistEntity rec) => EntityField rec typ -> Migration
```

The explicit `forall` here allows us to refer to `rec` in expressions in this function. Otherwise, we'll get a bizarre error like `No instance for PersistEntity rec0 found`.

Reply:
To explain the error you got:

The "true" first argument of `entityDef` is left polymorphic - you can pass any type constructor that has (as the last parameter) a type with a `PersistEntity` instance. So we can use `Nothing :: Maybe rec`, or `Left "hello" :: Either String rec`, or even `pure undefined :: IO rec`.

I imagine that you wrote `entityDef entityField`. This causes GHC to do the following inference:

```
-- terms:
entityDef   :: PersistEntity a => proxy a -> EntityDef
entityField :: EntityField rec typ

-- expression
(entityDef :: PersistEntity a => proxy a -> EntityDef) (entityField :: EntityField rec typ)

-- step 1) `entityDef` is a function, so we can accept a parameter.
-- step 2) let's type-check the first parameter to `entityDef` and verify that `entityField` works.
--         this means we'll try and unify the type of the first argument of the function and the type of the argument we provide

firstArgument 
    :: proxy a
entityField 
    :: EntityField rec typ

-- application is left associative, so we can add explicit parentheses:

firstArgument 
    :: (proxy) a
entityField 
    :: (EntityField rec) typ

-- we can unify `proxy ~ EntityField rec`. this is totally fine, just like `Either String rec` would have been.

-- step 3) substitute `proxy a ~ EntityField rec typ`:

entityDef 
    :: (PersistEntity a) => proxy a -> EntityDef
    -- substitute: `proxy ~ EntityField rec`, `a ~ typ`
    :: (PersistEntity typ) => EntityField rec typ -> EntityDef
```

Which is why GHC complained: No instance for `PersistEntity typ` !